velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,343 @@
+  void quickSort(
+      const PrefixSortIterator& start,
+      const PrefixSortIterator& end,
+      CompareMethod compare) const {
+    VELOX_CHECK(end >= start, "Invalid sort range.")
+    uint64_t len = end - start;
+
+    // Insertion sort on smallest arrays
+    if (len < kSmallSort) {
+      insertSort(start, end, compare);
+      return;
+    }
+
+    // Choose a partition element, v.
+
+    // Small arrays, median is the middle element.
+    auto m = start + len / 2;
+    if (len > kSmallSort) {
+      auto l = start;
+      auto n = end - 1;
+      // Big arrays, median is the pseudo-median of nine evenly spaced
+      // elements.
+      if (len > kMediumSort) {
+        uint64_t s = len / 8;
+        l = median3(l, l + s, l + 2 * s, compare);
+        m = median3(m - s, m, m + s, compare);
+        n = median3(n - 2 * s, n - s, n, compare);
+      }
+      // Mid-sized array, median is the median of the first, middle and last
+      // elements.
+      m = median3(l, m, n, compare);
+    }
+    auto a = start;
+    auto b = a;
+    auto c = end - 1;
+    auto d = c;
+
+    // Establish Invariant: v* (<v)* (>v)* v*,
+    // v* means elements that value equal to partition element v,
+    // (<v)* means elements that value little than partition element v,
+    // (>v)* means  elements that value bigger than partition element v.
+    while (true) {
+      int comparison;
+      while (b <= c && ((comparison = compare(*b, *m)) <= 0)) {
+        if (comparison == 0) {
+          if (a == m) {
+            m = b;
+          } else if (b == m) {
+            m = a;
+          }
+          swap(a++, b);
+        }
+        b++;
+      }
+      while (c >= b && ((comparison = compare(*c, *m)) >= 0)) {
+        if (comparison == 0) {
+          if (c == m) {
+            m = d;
+          } else if (d == m) {
+            m = c;
+          }
+          swap(c, d--);
+        }
+        c--;
+      }
+      if (b > c) {
+        break;
+      }
+      if (b == m) {
+        m = d;
+      }
+      // There is a useless assignment statement in the implementation of presto
+      // when c == m, remove it here,
+      // see: https://github.com/prestodb/presto/blob
+      // /542beb8d89fa86feb15008611f8900f0f316ef5f/
+      // presto-main/src/main/java/com/facebook/presto/operator
+      // /PagesIndexOrdering.java
+      swap(b++, c--);
+    }
+
+    // Swap partition elements back end middle
+    uint64_t s;
+    auto n = end;
+    s = std::min(a - start, b - a);
+    rangeSwap(start, b - s, s);
+    s = std::min(d - c, n - d - 1);
+    rangeSwap(b, n - s, s);
+
+    // Recursively sort non-partition-elements
+    s = b - a;
+    if (s > 1) {
+      quickSort(start, start + s, compare);
+    }
+    s = d - c;
+    if (s > 1) {
+      quickSort(n - s, n, compare);
+    }
+  }
+
+ private:
+  const uint64_t entrySize_;
+  BufferPtr swapBuffer_;
+  char* rawSwapBuffer_;
+  // Point to continuous sort data end(exclusive).
```

Comment:
Point ... -> Points just past the end of the data being sorted.

Do we need this member variable? Looks like we may remove it.